group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -829,20 +885,49 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
 
         // Get or create the member.
         if (memberId.isEmpty()) memberId = Uuid.randomUuid().toString();
-        final ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        ConsumerGroupMember member;
+        ConsumerGroupMember existingStaticMember = null;
+        if (instanceId == null) {
+            member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        } else {
+            existingStaticMember = group.staticMember(instanceId);
+            throwIfStaticMemberValidationFails(groupId, instanceId, existingStaticMember, memberEpoch, memberId);
+            member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        }
```

Comment:
I wonder if we could simplify it even more. For instance, would it be possible to have something like the following:

```
ConsumerGroupMember member;
ConsumerGroupMember updatedMember;

if (instanceId == null) {
  member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
  throwIfMemberEpochIsInvalid(member, memberEpoch, ownedTopicPartitions);
  log.info(...);
  updatedMember = new ConsumerGroupMember.Builder(member)
     ....
} else {
  // the new logic.
  // member is the current static member.
  // updatedMember is the updated current member or the new one.
}
```

Reply:
I have updated the logic in line with your suggestion.